test/unit/sdam/server_discovery_and_monitoring.prose.test.ts

```diff
@@ -0,0 +1,63 @@
+import { expect } from 'chai';
+import { createServer, type Server } from 'net';
+
+import { MongoClient, SERVER_HEARTBEAT_FAILED, SERVER_HEARTBEAT_STARTED } from '../../mongodb';
+
+describe('Heartbeat tests', function () {
+  let client: MongoClient;
+  // Mock server
+  let server: Server;
+  // Shared array
+  const events: string[] = [];
+
+  beforeEach(async function () {
+    // Create TCP server that responds to hellos by closing the connection
+    // and pushing "client connection created" to shared array
+    server = createServer(clientSocket => {
+      events.push('client connected');
+
+      clientSocket.once('data', () => {
+        events.push('client hello received');
+        clientSocket.end();
```

Comment:
```suggestion
        clientSocket.destroy();
```

This is a nit, and mostly opinion: if you aren't sending a final chunk and you aren't using the callback that `.end` accepts then I would reach for the API that "Ensures that no more I/O activity happens on this socket." vs the one that "Half-closes the socket. i.e., it sends a FIN packet. It is possible the server will still send some data."